### ISSUE TITLE
Distilling the Knowledge in a Neural Network

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#366]
https://arxiv.org/abs/1503.02531

**Comment by sweaterr** [comment]
Many insects have a larval form that is optimized for extracting energy and nutrients from the environment and a completely different adult form that is optimized for the very different requirements of traveling and reproduction.


In large-scale machine learning, we typically use very similar models for the training stage and the deployment stage despite their very different requirements: For tasks like speech and object recognition, training must extract structure from very large, highly redundant datasets but it does not need to operate in real time and it can use a huge amount of computation.


Deployment to a large number of users, however, has much more stringent requirements on latency and computational resources.


The analogy with insects suggests that we should be willing to train very cumbersome models if that makes it easier to extract structure from the data.


The cumbersome model could be an ensemble of separately trained models or a single very large model trained with a very strong regularizer such as dropout [9].


Once the cumbersome model has been trained, we can then use a different kind of training, which we call “distillation” to transfer the knowledge from the cumbersome model to a small model that is more suitable for deployment.


A version of this strategy has already been pioneered by Rich Caruana and his collaborators [1].


In their important paper they demonstrate convincingly that the knowledge acquired by a large ensemble of models can be transferred to a single small model.

**Comment by sweaterr** [comment]
A conceptual block that may have prevented more investigation of this very promising approach is that we tend to identify the knowledge in a trained model with the learned parameter values and this makes it hard to see how we can change the form of the model but keep the same knowledge.


A more abstract view of the knowledge, that frees it from any particular instantiation, is that it is a learned mapping from input vectors to output vectors.


For cumbersome models that learn to discriminate between a large number of classes, the normal training objective is to maximize the average log probability of the correct answer, but a side-effect of the learning is that the trained model assigns probabilities to all of the incorrect answers and even when these probabilities are very small, some of them are much larger than others.


The relative probabilities of incorrect answers tell us a lot about how the cumbersome model tends to generalize.


An image of a BMW, for example, may only have a very small chance of being mistaken for a garbage truck, but that mistake is still many times more probable than mistaking it for a carrot.


It is generally accepted that the objective function used for training should reflect the true objective of the user as closely as possible.


Despite this, models are usually trained to optimize performance on the training data when the real objective is to generalize well to new data.


It would clearly be better to train models to generalize well, but this requires information about the correct way to generalize and this information is not normally available.


When we are distilling the knowledge from a large model into a small one, however, we can train the small model to generalize in the same way as the large model.


If the cumbersome model generalizes well because, for example, it is the average of a large ensemble of different models, a small model trained to generalize in the same way will typically do much better on test data than a small model that is trained in the normal way on the same training set as was used to train the ensemble.


An obvious way to transfer the generalization ability of the cumbersome model to a small model is to use the class probabilities produced by the cumbersome model as “soft targets” for training the small model.


For this transfer stage, we could use the same training set or a separate “transfer” set.


When the cumbersome model is a large ensemble of simpler models, we can use an arithmetic or geometric mean of their individual predictive distributions as the soft targets.


When the soft targets have high entropy, they provide much more information per training case than hard targets and much less variance in the gradient between training cases, so the small model can often be trained on much less data than the original cumbersome model and using a much higher learning rate.


For tasks like MNIST in which the cumbersome model almost always produces the correct answer with very high confidence, much of the information about the learned function resides in the ratios of very small probabilities in the soft targets.


For example, one version of a 2 may be given a probability of 10−6 of being a 3 and 10−9 of being a 7 whereas for another version it may be the other way around.


This is valuable information that defines a rich similarity structure over the data (i.


e.


it says which 2’s look like 3’s and which look like 7’s) but it has very little influence on the cross-entropy cost function during the transfer stage because the probabilities are so close to zero.


Caruana and his collaborators circumvent this problem by using the logits (the inputs to the final softmax) rather than the probabilities produced by the softmax as the targets for learning the small model and they minimize the squared difference between the logits produced by the cumbersome model and the logits produced by the small model.


Our more general solution, called “distillation”, is to raise the temperature of the final softmax until the cumbersome model produces a suitably soft set of targets.


We then use the same high temperature when training the small model to match these soft targets.


We show later that matching the logits of the cumbersome model is actually a special case of distillation.


The transfer set that is used to train the small model could consist entirely of unlabeled data [1] or we could use the original training set.


We have found that using the original training set works well, especially if we add a small term to the objective function that encourages the small model to predict the true targets as well as matching the soft targets provided by the cumbersome model.


Typically, the small model cannot exactly match the soft targets and erring in the direction of the correct answer turns out to be helpful.

**Comment by sweaterr** [comment]
When the correct labels are known for all or some of the transfer set, this method can be significantly improved by also training the distilled model to produce the correct labels.
올바른 라벨이 전 송 세트 전체 또는 일부에 알려져있을 때,이 방법은 증류 된 모델을 교육하여 올바른 라벨을 생성함으로써 현저히 향상시킬 수 있습니다.

One way to do this is to use the correct labels to modify the soft targets, but we found that a better way is to simply use a weighted average of two different objective functions.
이를 수행하는 한 가지 방법은 올바른 라벨을 사용하여 부드러운 타겟을 수정하는 것이지만 더 나은 방법은 두 가지 목적 함수의 가중 평균을 사용하는 것입니다.

The first objective function is the cross entropy with the soft targets and this cross entropy is computed using the same high temperature in the softmax of the distilled model as was used for generating the soft targets from the cumbersome model.
첫 번째 목적 함수는 부드러운 표적과의 교차 엔트로피이며이 크로스 엔트로피는 성가신 모델에서 부드러운 표적을 생성하는 데 사용 된 것과 같이 증류 된 모델의 softmax에서 동일한 고온을 사용하여 계산됩니다.

The second objective function is the cross entropy with the correct labels.
두 번째 목적 함수는 올바른 레이블을 가진 교차 엔트로피입니다.

This is computed using exactly the same logits in softmax of the distilled model but at a temperature of 1.
이것은 증류 된 모델의 softmax에서 정확히 동일한 로지를 사용하지만 온도가 1 인 경우 계산됩니다.

We found that the best results were generally obtained by using a condiderably lower weight on the second objective function.
우리는 최상의 결과가 일반적으로 두 번째 목적 함수에 비할 때 더 낮은 가중치를 사용함으로써 얻어 짐을 발견했습니다.

Since the magnitudes of the gradients produced by the soft targets scale as 1/T 2 it is important to multiply them by T 2 when using both hard and soft targets.
소프트 타겟에 의해 생성 된 그래디언트의 크기는 1 / T2로 스케일링되기 때문에 하드 타겟과 소프트 타겟을 모두 사용할 때 T2를 곱하는 것이 중요합니다.

This ensures that the relative contributions of the hard and soft targets remain roughly unchanged if the temperature used for distillation is changed while experimenting with meta-parameters.
메타 매개 변수를 실험하면서 증류에 사용 된 온도가 변경되면 하드 타겟 및 소프트 타겟의 상대 기여도가 거의 변하지 않습니다.

**Comment by sweaterr** [comment]
## 2.1 Matching logits is a special case of distillation

**Comment by sweaterr** [comment]
## 3 Preliminary experiments on MNIST

To see how well distillation works, we trained a single large neural net with two hidden layers
of 1200 rectified linear hidden units on all 60,000 training cases. The net was strongly regularized
using dropout and weight-constraints as described in [5]. Dropout can be viewed as a way of training
an exponentially large ensemble of models that share weights. In addition, the input images were
jittered by up to two pixels in any direction. This net achieved 67 test errors whereas a smaller
net with two hidden layers of 800 rectified linear hidden units and no regularization achieved 146
errors. But if the smaller net was regularized solely by adding the additional task of matching the soft
targets produced by the large net at a temperature of 20, it achieved 74 test errors. This shows that
soft targets can transfer a great deal of knowledge to the distilled model, including the knowledge
about how to generalize that is learned from translated training data even though the transfer set does
not contain any translations.
When the distilled net had 300 or more units in each of its two hidden layers, all temperatures above
8 gave fairly similar results. But when this was radically reduced to 30 units per layer, temperatures
in the range 2.5 to 4 worked significantly better than higher or lower temperatures.
We then tried omitting all examples of the digit 3 from the transfer set. So from the perspective
of the distilled model, 3 is a mythical digit that it has never seen. Despite this, the distilled model
only makes 206 test errors of which 133 are on the 1010 threes in the test set. Most of the errors
are caused by the fact that the learned bias for the 3 class is much too low. If this bias is increased
by 3.5 (which optimizes overall performance on the test set), the distilled model makes 109 errors
of which 14 are on 3s. So with the right bias, the distilled model gets 98.6% of the test 3s correct
despite never having seen a 3 during training. If the transfer set contains only the 7s and 8s from the
training set, the distilled model makes 47.3% test errors, but when the biases for 7 and 8 are reduced
by 7.6 to optimize test performance, this falls to 13.2% test errors.

**Comment by sweaterr** [comment]
## 5 Training ensembles of specialists on very big datasets

Training an ensemble of models is a very simple way to take advantage of parallel computation and the usual objection that an ensemble requires too much computation at test time can be dealt with by using distillation.
앙상블 모델을 훈련하는 것은 병렬 계산을 이용하는 매우 간단한 방법이며, 앙상블이 테스트 시간에 너무 많은 계산을 요구하는 일반적인 반대는 증류를 사용하여 처리 할 수 ​​있습니다.

There is, however, another important objection to ensembles: If the individual models are large neural networks and the dataset is very large, the amount of computation required at training time is excessive, even though it is easy to parallelize.
그러나 앙상블에 또 다른 중요한 반대가 있습니다. 개별 모델이 대형 신경 네트워크이고 데이터 세트가 매우 큰 경우 병렬화하기 쉽지만 교육 시간에 필요한 계산량이 과도합니다.

In this section we give an example of such a dataset and we show how learning specialist models that each focus on a different confusable subset of the classes can reduce the total amount of computation required to learn an ensemble.
이 섹션에서는 그러한 데이터 세트의 예를 제시하고 클래스의 다른 혼동 가능한 하위 세트에 중점을 두는 학습 전문가 모델이 앙상블을 학습하는 데 필요한 총 계산량을 줄이는 방법을 보여줍니다.

The main problem with specialists that focus on making fine-grained distinctions is that they overfit very easily and we describe how this overfitting may be prevented by using soft targets.
세밀한 구분에 초점을 맞춘 전문가의 주된 문제는 매우 쉽게 지나치게 적합하다는 것인데, 우리는이 과핑이 어떻게 부드러운 타겟을 사용하여 방지 될 수 있는지 설명합니다.

**Comment by sweaterr** [comment]
## 5.1 The JFT dataset
![image](https://user-images.githubusercontent.com/1518919/40064855-5c31c89e-589b-11e8-81d9-15da621ceb5f.png)

JFT is an internal Google dataset that has 100 million labeled images with 15,000 labels.
JFT는 15,000 개의 라벨이있는 1 억 개의 라벨 이미지가있는 내부 Google 데이터 세트입니다.

When we did this work, Google’s baseline model for JFT was a deep convolutional neural network [7] that had been trained for about six months using asynchronous stochastic gradient descent on a large number of cores.
이 작업을 수행했을 때 Google의 JFT에 대한 기본 모델은 많은 수의 코어에서 비동기 확률 적 그래디언트 디센트를 사용하여 약 6 개월 동안 훈련 된 심층 컨볼 루션 신경망이었습니다 [7].

This training used two types of parallelism [2].
이 훈련은 두 가지 유형의 병렬 처리를 사용했습니다 [2].

First, there were many replicas of the neural net running on different sets of cores and processing different mini-batches from the training set.
첫째, 신경망의 여러 복제본이 서로 다른 코어 세트에서 실행되며 교육 세트와는 다른 미니 배치를 처리합니다.

Each replica computes the average gradient on its current mini-batch and sends this gradient to a sharded parameter server which sends back new values for the parameters.
각 복제본은 현재 미니 배치의 평균 그래디언트를 계산하고이 그래디언트를 매개 변수에 새 값을 다시 보내는 샤드 된 매개 변수 서버로 보냅니다.

These new values reflect all of the gradients received by the parameter server since the last time it sent parameters to the replica.
이 새 값은 매개 변수 서버가 매개 변수를 복제본으로 마지막으로 보낸 이후로 수신 한 모든 그라디언트를 반영합니다.

Second, each replica is spread over multiple cores by putting different subsets of the neurons on each core.
둘째, 각각의 복제본은 각 코어에 뉴런의 다른 부분 집합을 배치하여 여러 개의 코어에 분산됩니다.

Ensemble training is yet a third type of parallelism that can be wrapped around the other two types, but only if a lot more cores are available.
Ensemble 교육은 다른 두 가지 유형으로 둘러 쌀 수있는 세 번째 유형의 병렬 처리이지만 훨씬 많은 코어를 사용할 수있는 경우에만 가능합니다.

Waiting for several years to train an ensemble of models was not an option, so we needed a much faster way to improve the baseline model.
모델 앙상블을 수련하기 위해 수년을 기다리는 것은 선택 사항이 아니므로 기본 모델을 개선하는 훨씬 빠른 방법이 필요했습니다.

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/40069312-de3c6bfa-58a5-11e8-996c-47f2e91a4efb.png)

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/40070351-ca7a1556-58a8-11e8-8d07-3c520155d2a6.png)

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/40070367-d2b8faa2-58a8-11e8-96a6-7c0724699313.png)
![image](https://user-images.githubusercontent.com/1518919/40070376-d7f21256-58a8-11e8-8b67-aedbf27d482a.png)

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/40070381-ddc28b66-58a8-11e8-8f73-401c956fd07d.png)

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/40070390-e45aef4a-58a8-11e8-92a3-e23ad30cd9f9.png)

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/40070402-eb560564-58a8-11e8-9a7c-eba3e70a061e.png)
